File: go/word-count/word_count.go
// Package wordcount contains a solution to the Exercism word-count exercise.
package wordcount

import (
	"regexp"
	"strings"
)

// Frequency is the frequency that words appears in a phrase.
type Frequency map[string]int

// WordCount counts the number of words within a given phrase, returning a
// Frequency value.
func WordCount(phrase string) Frequency {
	re := regexp.MustCompile(`[^a-zA-Z0-9']`)
	frequency := make(Frequency)

	p := re.ReplaceAllString(strings.ToLower(phrase), " ")

	for _, word := range strings.Fields(p) {
		word = strings.Trim(word, `'`)
		frequency[word]++
	}

	return frequency
}

File: ruby/word-count/word_count.rb
class Phrase
  def initialize(phrase)
    @phrase = phrase
  end

  attr_reader :phrase

  def words
    phrase.downcase.scan(/\b[\w']+\b/)
  end

  def word_count()
    words.each_with_object(Hash.new(0)) do |word, count|
      count[word] += 1
    end
  end
end
